pandas/core/index.py

```diff
@@ -2510,7 +2527,342 @@ def invalid_op(self, other=None):
+        result = object.__new__(cls)
+
+        if not isinstance(values, ABCCategorical):
+            from pandas.core.categorical import Categorical
+            values = Categorical(values, categories=categories, ordered=True)
+        elif categories is not None:
+            values = values.set_categories(categories)
+
+        result._data = values
+        result.name = name
+        for k, v in compat.iteritems(kwargs):
+            setattr(result,k,v)
+
+        result._reset_identity()
+        return result
+
+    def equals(self, other):
+        """
+        Determines if two CategorialIndex objects contain the same elements.
+        """
+        if self.is_(other):
+            return True
+
+        try:
+            if is_categorical_dtype(other):
+                if not self.categories.equals(other.categories):
+                    return False
+            else:
+                from pandas import Categorical
+                other = Categorical(other, categories=self.categories, ordered=True)
+                if isnull(other).any():
+                    return False
+
+            return (self._data == other).all()
+        except:
+            pass
+
+        return False
+
+    @property
+    def inferred_type(self):
+        return 'categorical'
+
+    @property
+    def values(self):
+        """ return the underlying data, which is a Categorical """
+        return self._data
+
+    @property
+    def codes(self):
+        return self._data.codes
+
+    @property
+    def categories(self):
+        return self._data.categories
+
+    @property
+    def ordered(self):
+        return self._data.ordered
+
+    def __contains__(self, key):
+        hash(key)
+        return key in self.categories
+
+    def __array__(self, result=None):
+        """ the array interface, return my values """
+        return np.array(self._data)
+
+    def _array_values(self):
+        return self.values
+
+    def argsort(self, *args, **kwargs):
+        return self.values.argsort(*args, **kwargs)
+
+    @cache_readonly
+    def _engine(self):
+
+        # we are going to look things up with the codes themselves
+        return self._engine_type(lambda: self.codes.astype('i8'), len(self))
+
+    @cache_readonly
+    def is_unique(self):
+        return not self.duplicated().any()
+
+    @Appender(_shared_docs['duplicated'] % _index_doc_kwargs)
+    def duplicated(self, take_last=False):
+        from pandas.hashtable import duplicated_int64
+        return duplicated_int64(self.codes.astype('i8'), take_last)
+
+    def get_loc(self, key, method=None):
+        """
+        Get integer location for requested label
 
+        Parameters
+        ----------
+        key : label
+        method : {None}
+            * default: exact matches only.
+
+        Returns
+        -------
+        loc : int if unique index, possibly slice or mask if not
+        """
+        tcodes = self.categories.get_indexer([key])
```

Comment:
any particular reason why you can't use `self.categories.get_loc`? That seems a little more robust, especially to the possibility of `key` appearing multiple times in the index.